Mathlib/Algebra/Category/AlgebraCat/Basic.lean

```diff
@@ -46,22 +46,88 @@ instance : CoeSort (AlgebraCat R) (Type v) :=
 
 attribute [coe] AlgebraCat.carrier
 
+/-- The object in the category of R-algebras associated to a type equipped with the appropriate
+typeclasses. -/
+def of (X : Type v) [Ring X] [Algebra R X] : AlgebraCat.{v} R :=
+  ⟨X⟩
+
+@[simp]
+lemma coe_of (X : Type v) [Ring X] [Algebra R X] : (of R X : Type v) = X :=
+  rfl
+
+variable {R} in
+/-- The type of morphisms in `AlgebraCat R`. -/
+@[ext]
+structure Hom (A B : AlgebraCat.{v} R) where
```

Comment:
Similarly,
```suggestion
structure Hom (A B : AlgebraCat.{v} R) where ofHom ::
```
might let you drop `ofHom` below.

Reply:
This does seem to produce `Hom.ofHom` which is undesirable. Tricking it with `_root_.AlgebraCat.ofHom ::` also does not work.

Reply:
The `export` command allows you to remove the `Hom` namespace, something like:
```lean
namespace AlgebraCat

export Hom (hom)

end AlgebraCat
```

Reply:
Good point!

Note for myself to check: `(of A : Type*)` does not unify with `A` at reducible transparency, right?

Reply:
Indeed:

```
example {R A : Type u} [CommRing R] [Ring A] [Algebra R A] : (of R A : Type u) = A := by
  fail_if_success with_reducible rfl -- Not reducibly defeq
  rfl -- But they are defeq.
```

Reply:
Should we make `AlgebraCat.of` an `abbrev` then? This would fix the `ext` issue.

Reply:
In my mind, API boundaries in category theory should be a semipermeable membrane where we easily flow into the concrete types themselves but it is hard to flow from them to objects and morphisms. So I think an `abbrev` is reasonable.

Reply:
That might be a good idea, because then `↑(of R X) = ↑⟨X⟩ = X` is a (reducible) definitional equality. I'll see if that causes any downstream breakage...

Reply:
It does conflict with Eric's suggestion above though. In theory, if `of` is actually the constructor, then it should be eta reduced away at reducible transparency.

Reply:
And it allows for a lot of proofs to be entirely automated! Feel free to cherry-pick commit [`250bef56`](https://github.com/leanprover-community/mathlib4/commit/250bef56683a482ea81c4024f54910df376a5bee) :)

Reply:
I raced you :) but you spotted some more golfs

Reply:
> It does conflict with Eric's suggestion above though. In theory, if `of` is actually the constructor, then it should be eta reduced away at reducible transparency.

Exactly, so if we can't figure out how to make `R` explicit in `mk` then `abbrev of` would be the alternative.

Reply:
Should `ofHom` also get the same treatment?

Reply:
Ideally we could do something like
```lean
structure Hom' {R : Type*} [CommRing R] (A : Type u) (B : Type v) [Ring A] [Ring B] [Algebra R A]
    [Algebra R B] where
  hom : A →ₐ[R] B 

instance : Category (AlgebraCat.{v} R) where 
  Hom := fun ⟨A⟩ ⟨B⟩ => Hom A B
  id := fun ⟨A⟩ => ⟨AlgHom.id R A⟩
  comp := fun ⟨f⟩ ⟨g⟩ => ⟨g.comp f⟩
```
but Lean cannot find the ring instance on `A` and `B` when destructuring

Reply:
I don't think we should be using pattern matches here, as they generate worse terms.

Reply:
In my experience, the support for `match` reduction exceeds that of eta expansions.

Reply:
See [Zulip](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/optimal.20instance.20design.20for.20structure.20wrappers) for an example. Feel free to disabuse me of this belief.

Reply:
Here